### PR TITLE
[IRGen] Generate proper type descriptors for ObjC subscript accessors

### PR DESCRIPTION
**Pull request by jrose-apple**
[SR-8356](https://bugs.swift.org/browse/SR-8356) / rdar://problem/42284266

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

File: lib/IRGen/GenDecl.cpp [hunk 1]
```diff
@@ -437,7 +437,8 @@ class ObjCProtocolInitializerVisitor
     Builder.CreateCall(protocol_addMethodDescription, getterArgs);
     
     if (prop->isSettable(nullptr)) {
-      emitObjCSetterDescriptorParts(IGM, prop, name, types, imp);
+      emitObjCSetterDescriptorParts(IGM, prop, /*extended*/true,
+                                    name, types, imp);
```
[the remark this answers is not in the document]
Comment (jrose-apple) [reply lib/IRGen/GenDecl.cpp]: Protocols do, non-protocols don't. I don't know the story further than that, though; I just imitated what was already there. (That's why I tagged Greg for review, really.)

[the remark this answers is not in the document]
Comment (rjmccall) [reply lib/IRGen/GenDecl.cpp]: Ah, I guess that makes as much sense as anything.  Could you make the fact that that's the reason behind the decision more explicit in the source?

[the remark this answers is not in the document]
Comment (jrose-apple) [reply lib/IRGen/GenDecl.cpp]: Okay, I don't think this is correct after all. There's already a separate path for getting extended type encodings for protocols. @gparker42, how is this supposed to work?

**Comment by jrose-apple** [comment]
\*sigh\* ObjCBool…

**Review by rjmccall — COMMENTED** [review]
This seems reasonable to me.  What's going on with `ObjCBool`?

**Comment by jrose-apple** [comment]
Oh, just that the test case I modified was already limited to `CPU=x86_64` because whoever wrote it originally didn't feel like conditionalizing it, but I forgot that that includes the 64-bit simulator. I think it's probably okay to just change it to OS=macosx, since it wasn't going to be testing most platforms anyway.

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Pull request merged**